master/buildbot/util/git.py

```diff
@@ -270,53 +345,344 @@ def checkFeatureSupport(self):
+        )
+        return self.step.master
+
+    @defer.inlineCallbacks
+    def _download_file(
+        self, path: str, content: str, mode: int, workdir: str | None = None,
+    ) -> Generator[defer.Deferred[Any], Any, None]:
+        assert isinstance(self.step, CompositeStepMixin)
+        yield self.step.downloadFileContentToWorker(
+            path, content, mode=mode, workdir=workdir,
+        )
+
+    @defer.inlineCallbacks
+    def download_auth_files_if_needed(
+        self,
+        workdir: str,
+        download_wrapper_script: bool = False,
+    ) -> Generator[defer.Deferred[Any], Any, int]:
+        if self.ssh_private_key is None:
             return RC_SUCCESS
 
-        yield self.runRmdir(self._getSshDataPath())
+        assert (
+            isinstance(self.step, CompositeStepMixin) and
+            isinstance(self.step, GitMixin)
+        )
+
+        workdir = self._get_auth_data_path(workdir)
+        yield self.step.runMkdir(workdir)
+
+        return_code = yield super().download_auth_files_if_needed(
+            workdir=workdir,
+            download_wrapper_script=(
+                download_wrapper_script or
+                not self.step.supportsSshPrivateKeyAsEnvOption
+            ),
+        )
+        return return_code
+
+    @defer.inlineCallbacks
+    def remove_auth_files_if_needed(
+        self, workdir: str,
+    ) -> Generator[defer.Deferred[Any], Any, int]:
+        if not self.did_download_auth_files:
+            return RC_SUCCESS
+
+        assert isinstance(self.step, CompositeStepMixin)
+        yield self.step.runRmdir(self._get_auth_data_path(workdir))
+        return RC_SUCCESS
+
+
+class GitServiceAuth(AbstractGitAuth):
+
+    def __init__(
+        self,
+        service: BuildbotService,
+        ssh_private_key: IRenderable | None = None,
+        ssh_host_key: IRenderable | None = None,
+        ssh_known_hosts: IRenderable | None = None,
+    ) -> None:
+        self._service = service
+
+        super().__init__(ssh_private_key, ssh_host_key, ssh_known_hosts)
+
+    @property
+    def _path_module(self):
+        return os.path
+
+    @property
+    def _master(self):
+        assert self._service.master is not None
+        return self._service.master
+
+    @defer.inlineCallbacks
+    def _download_file(
+        self, path: str, content: str, mode: int, workdir: str | None = None,
+    ) -> Generator[defer.Deferred[Any], Any, None]:
+        writeLocalFile(path, content, mode=mode)
+        yield defer.succeed(None)
+
+    @defer.inlineCallbacks
+    def remove_auth_files_if_needed(
+        self, workdir: str,
+    ) -> Generator[defer.Deferred[Any], Any, int]:
+        if not self.did_download_auth_files:
+            return RC_SUCCESS
+
+        Path(
+            self._get_ssh_private_key_path(workdir)
+        ).unlink(missing_ok=True)
+        Path(
+            self._get_ssh_host_key_path(workdir)
+        ).unlink(missing_ok=True)
+
+        yield defer.succeed(None)
```

Comment:
This doesn't do anything.

Reply:
I should have added a comment.
This is awkward stuff needed if we want to type the function in combination to of using `inlineCallback` (one more case were porting to async as mentioned in #7371 would be helpful).

Since `GitStepAuth` implementation need to return a `Deferred` and `inlineCallback` make using multiple far more readable, we need to use it in `AbstractGitAuth`.
Then, as `AbstractGitAuth` is tagged as `inlineCallback`, the function's return type need to be `Generator[...]`.
Which make `GitServiceAuth` need to return a `Generator` as well as be an `inlineCallback`, thus the weird useless yield...

As I said, I typing of `inlineCallback` function's return in a separate commit that can be dropped as it probably bring more confusion that help in anything.

Reply:
Maybe it makes sense to not use typing until we migrate to async/await? Right now typing adds a lot of noise to inlineCallbacks function declarations and also requires cumbersome code.